src/dump.complex.js

```diff
@@ -0,0 +1,100 @@
+'use strict'
+
+const dumpBool = require('./dump.bool')
+const dumpFloat = require('./dump.float')
+const dumpInt = require('./dump.int')
+const dumpNull = require('./dump.null')
+const dumpString = require('./dump.string')
+const dumpSymbol = require('./dump.symbol')
+const dumpSymbolLink = require('./dump.symbol.link')
+
+/**
+ * @param {string} symbol
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const dumpSymbolOrLink = (symbol, symcache) => {
+  const pos = symcache.indexOf(symbol)
+  if (pos === -1) {
+    symcache.push(symbol)
+    return dumpSymbol(symbol)
+  }
+
+  return dumpSymbolLink(pos)
+}
+
+/**
+ * @param {Array} arr
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const dumpArray = (arr, symcache) => {
+  const head = Buffer.from('[', 'utf8')
+  const len = dumpInt(arr.length).slice(1)
+  const data = []
+
+  for (const val of arr) {
+    data.push(encodeType(val, symcache))
+  }
+
+  return Buffer.concat([
+    head,
+    len,
+    ...data
+  ])
+}
+
+/**
+ * @param {Object} obj
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const dumpHash = (obj, symcache) => {
+  const head = Buffer.from('{', 'utf8')
+  const len = dumpInt(Object.keys(obj).length).slice(1)
+
+  const data = []
+  for (const key in obj) {
+    data.push(dumpSymbolOrLink(key, symcache), encodeType(obj[key], symcache))
+  }
+
+  return Buffer.concat([
+    head,
+    len,
+    ...data
+  ])
+}
+
+/**
+ * @param {any} val
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const encodeType = (val, symcache) => { // see http://jakegoulding.com/blog/2013/01/15/a-little-dip-into-rubys-marshal-format
+  let symbol
+
+  switch (typeof val) {
+    case 'string': return dumpString(val)
+    case 'boolean': return dumpBool(val)
+    case 'number':
+      if (Number.isInteger(val)) return dumpInt(val)
+      return dumpFloat(val)
+    case 'object':
+      if (val === null) return dumpNull()
+      if (Array.isArray(val)) return dumpArray(val, symcache)
+      return dumpHash(val, symcache)
+    case 'symbol':
+      symbol = val.toString().replace('Symbol(', '').replace(')', '')
+      return dumpSymbolOrLink(symbol, symcache)
+    case 'undefined': // undefined or null is nil in ruby
+      return dumpNull()
```

Comment:
its not idempotent:

```js
'use strict'

const { inspect } = require('util')
const marshal = require('../index')


const test = { foo: undefined }
const serialized1 = marshal.dump(test)

const s2 = marshal.load(serialized1)
console.log(s2) // { foo: null }
```

Reply:
expected: `{ foo: undefined }`

Reply:
(important when parsing json)

Reply:
how does `Oj` handle this?

Reply:
the json spec just knows about "null": https://www.json.org/json-en.html

Reply:
maybe we should not support `undefined` and go with `ERR_MARSHAL_TYPE_NOT_SUPPORTED` in those cases?

Reply:
the problem is that ruby doesn't have undefined type but nil in js means null or undefined usually, that's why I thought to parse it as null, should we throw error in case of undefined?

Reply:
i think we should throw for undefined, else we run into issues when serializing back and forth